docs/src/topics/assembly.md

```diff
@@ -83,3 +84,302 @@ for t in 1:timesteps
+    end
+    return
+end
+nothing # hide
+```
+
+The core operation, `K[I, J] += Ke[i, j]`, can still be broken down into three equivalent
+steps:
+
+```julia
+tmp1 = K[I, J]
+tmp2 = tmp1 + Ke[i, j]
+K[I, J] = tmp2
+```
+
+The key difference here is that we index using integers (`I`, `J`, `i`, and `j`) which means
+that `tmp1` and `tmp2` are scalars which don't need to be allocated on the heap. This
+stragety thus eliminates all allocations that were present in the first strategy. However,
+we still lookup the same location in `K` twice, and we still have a random access pattern.
+
+#### Strategy 3: scalar indexing with single lookup
+
+To improve on the second strategy we will get rid of the double lookup into the sparse
+matrix `K`. While Julia doesn't have a "`+=`"-operation there is a `addindex!`-function in
+Ferrite which does exactly what we want: it adds a value to a specific location in a sparse
+matrix using a single lookup.
+
+```@example assembly-perf
+function assemble_v3(_, K, dofs, Ke)
+    for (i, I) in pairs(dofs)
+        for (j, J) in pairs(dofs)
+            Ferrite.addindex!(K, Ke[i, j], I, J)
+        end
+    end
+    return
+end
+nothing # hide
+```
+
+With this method we remove the double lookup, but the issue of random access patterns still
+remains.
+
+#### Strategy 4: using an assembler
+
+Finally, the last strategy we consider uses an assembler. The assembler is a specific
+datastructure that pre-allocates some workspace to make the assembly more efficient:
+
+```@example assembly-perf
+function assemble_v4(assembler, _, dofs, Ke)
+    assemble!(assembler, dofs, Ke)
+    return
+end
+nothing # hide
+```
+
+The extra workspace inside the assembler is used to sort the dofs when `assemble!` is
+called. After sorting it is possible to loop over the sparse matrix data structure and
+insert all elements of `Ke` in one go instead of having to lookup locations randomly.
+
+### Single element assembly
+
+First we will compare the four functions above for a single assembly operation, i.e.
+inserting one local matrix into the global matrix. For this we simply create a random local
+matrix since we are not conserned with the actual values. We also pick the "middle" element
+and extract the dofs for that element. Finally, an assembler is created with
+`start_assemble` to use with the fourth strategy.
+
+```@example assembly-perf
+dofs_per_cell = ndofs_per_cell(dh)
+const Ke = rand(dofs_per_cell, dofs_per_cell)
+const dofs = celldofs(dh, N * N ÷ 2)
+
+const assembler = start_assemble(K)
+nothing # hide
+```
+
+We use BenchmarkTools to measure the performance:
+
+```@example assembly-perf
+assemble_v1(assembler, K, dofs, Ke) # hide
+assemble_v2(assembler, K, dofs, Ke) # hide
+assemble_v3(assembler, K, dofs, Ke) # hide
+assemble_v4(assembler, K, dofs, Ke) # hide
+nothing                             # hide
+```
+
+```julia
+using BenchmarkTools
+
+@btime assemble_v1(assembler, K, dofs, Ke) evals = 10 setup = Ferrite.fillzero!(K)
```

Comment:
Any reason why the regular `fill!` cannot be used? If so, a small remark should probably be made? 
```suggestion
@btime assemble_v1(assembler, K, dofs, Ke) evals = 10 setup = fill!(K, 0)
```